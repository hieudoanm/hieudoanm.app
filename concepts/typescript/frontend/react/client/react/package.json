{
  "name": "@react/react",
  "version": "0.0.1",
  "description": "React - React",
  "main": "index.js",
  "repository": "git@github.com:hieudoanm/hieudoanm.git",
  "author": "hieudoanm <hieumdoan@gmail.com>",
  "license": "GPL-3.0",
  "private": true,
  "type": "module",
  "scripts": {
    "dev": "vite",
    "build": "tsc && vite build",
    "preview": "vite preview",
    "test": "jest",
    "lint": "eslint . --fix --ext .ts,.tsx"
  },
  "dependencies": {
    "@emotion/react": "11.11.3",
    "@emotion/styled": "11.11.0",
    "@fontsource/roboto": "5.0.8",
    "@mui/icons-material": "5.15.7",
    "@mui/material": "5.15.7",
    "@types/node": "20.11.13",
    "apexcharts": "3.45.2",
    "pino": "8.17.2",
    "react": "18.2.0",
    "react-apexcharts": "1.4.1",
    "react-dom": "18.2.0",
    "react-router-dom": "6.21.3"
  },
  "devDependencies": {
    "@testing-library/jest-dom": "6.4.1",
    "@testing-library/react": "14.2.0",
    "@types/jest": "29.5.11",
    "@types/react": "18.2.48",
    "@types/react-dom": "18.2.18",
    "@typescript-eslint/eslint-plugin": "6.20.0",
    "@typescript-eslint/parser": "6.20.0",
    "@vitejs/plugin-react-swc": "3.6.0",
    "autoprefixer": "10.4.17",
    "eslint": "8.56.0",
    "eslint-plugin-react": "7.33.2",
    "jest": "29.7.0",
    "jest-environment-jsdom": "29.7.0",
    "kill-port": "2.0.1",
    "postcss": "8.4.33",
    "sass": "1.70.0",
    "tailwindcss": "3.4.1",
    "ts-jest": "29.1.2",
    "ts-node": "10.9.2",
    "typescript": "5.3.3",
    "vite": "5.0.12"
  },
  "keywords": [
    "react"
  ]
}
