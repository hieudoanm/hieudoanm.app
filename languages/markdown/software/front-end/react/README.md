---
title: 'React'
date: '2021-12-12'
author: 'Hieu Doan'
description: 'React'
---

### Class Component vs. Functional Component

- Difference syntax
- Before the introduction of hooks, functional components were stateless. However, with React 16.8, you can implement states with the useState hook to create a stateful component (just like the class component).
- Usage
  - Functional components with hooks are concise and more straightforward to code with. They perform exactly as the class component; this implies no difference between the two other than syntax.
  - By using just functional components in your project, you drastically eliminate the need to refactor the class component into a functional component when it grows.
  - Since classes confuse both people and machines, most especially the this keyword, you don't have to worry about this anymore in functional components.
  - No need for unnecessary method binding like we always do in the class component.
  - Sharing stateful logic between components is tedious in a class-based approach.

### `import`

- After compiling, the code we actually get is `return React.createElement`
