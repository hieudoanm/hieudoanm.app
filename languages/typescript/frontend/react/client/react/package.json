{
  "name": "@react/react",
  "version": "0.0.1",
  "description": "React - React",
  "main": "index.js",
  "repository": "git@github.com:houseoftypescript/web-react-template.git",
  "author": "hieudoanm <hieumdoan@gmail.com>",
  "license": "GPL-3.0",
  "private": true,
  "type": "module",
  "scripts": {
    "dev": "vite",
    "build": "tsc && vite build",
    "preview": "vite preview",
    "test": "jest",
    "lint": "eslint . --fix --ext .ts,.tsx"
  },
  "dependencies": {
    "@emotion/react": "11.11.1",
    "@emotion/styled": "11.11.0",
    "@fontsource/roboto": "5.0.8",
    "@mui/icons-material": "5.14.18",
    "@mui/material": "5.14.18",
    "@types/node": "20.9.1",
    "apexcharts": "3.44.0",
    "pino": "8.16.2",
    "react": "18.2.0",
    "react-apexcharts": "1.4.1",
    "react-dom": "18.2.0",
    "react-router-dom": "6.19.0"
  },
  "devDependencies": {
    "@testing-library/jest-dom": "6.1.4",
    "@testing-library/react": "14.1.2",
    "@types/jest": "29.5.8",
    "@types/react": "18.2.37",
    "@types/react-dom": "18.2.15",
    "@typescript-eslint/eslint-plugin": "6.11.0",
    "@typescript-eslint/parser": "6.11.0",
    "@vitejs/plugin-react-swc": "3.5.0",
    "autoprefixer": "10.4.16",
    "eslint": "8.54.0",
    "eslint-plugin-react": "7.33.2",
    "jest": "29.7.0",
    "jest-environment-jsdom": "29.7.0",
    "kill-port": "2.0.1",
    "postcss": "8.4.31",
    "sass": "1.69.5",
    "tailwindcss": "3.3.5",
    "ts-jest": "29.1.1",
    "ts-node": "10.9.1",
    "typescript": "5.2.2",
    "vite": "5.0.0"
  },
  "keywords": [
    "react"
  ]
}
