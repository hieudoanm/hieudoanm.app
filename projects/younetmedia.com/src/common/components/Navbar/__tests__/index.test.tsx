import { render } from '@testing-library/react';
import { Navbar } from '../index';

jest.mock('next/router', () => {
  return { useRouter: jest.fn() };
});

describe('Navbar', () => {
  it('renders correctly', () => {
    const { container } = render(<Navbar />);
    expect(container).toMatchSnapshot();
  });
});
