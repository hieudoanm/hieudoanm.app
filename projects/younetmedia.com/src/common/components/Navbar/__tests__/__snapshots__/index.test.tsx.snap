// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctly 1`] = `
<div>
  <nav
    class="border-b shadow"
  >
    <div
      class="container mx-auto px-8 py-4"
    >
      <div
        class="flex items-center justify-between"
      >
        <div
          class="flex items-center gap-x-2"
        >
          <a
            href="/"
          >
            <h1
              class="uppercase"
            >
              Buzz Metrics
            </h1>
          </a>
          <a
            href="/socialheat"
          >
            <p>
              Social Heat
            </p>
          </a>
          <a
            href="/socialtrend"
          >
            <p>
              Social Trend
            </p>
          </a>
        </div>
        <div>
          <button
            class="btn bg-blue-500 text-white uppercase"
            type="button"
          >
            Log Out
          </button>
        </div>
      </div>
    </div>
  </nav>
</div>
`;
